src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2568,7 +2568,8 @@ module Crystal
       # Try to resolve the sizeof right now to a number literal
       # (useful for sizeof inside as a generic type argument, but also
       # to make it easier for LLVM to optimize things)
-      if type && !node.exp.is_a?(TypeOf)
+      if type && !node.exp.is_a?(TypeOf) &&
+         !(type.module? || (type.abstract? && type.struct?))
```

Comment:
Can't this be moved upwards together with the check of `type.is_a?(GenericType)`? I thought the notion of an instantiable type was already defined somewhere.

Reply:
Here in `MainVisitor` we compute it if we can. Later in `CleanupTransformer` we check that it's only a class. So, doing `sizeof` an abstract struct or module can be done and works, just not eagerly.

To be honest, now I'm not convinced that `sizeof` should work eagerly at all. I did that so one can do, for example, `StaticArray(Int32, sizeof(Int32))`, but that's a pretty contrived example. But for now I'm just fixing broken things, not changing the semantics of the language or introducing breaking changes. But eventually we can define the semantics we want for this and fix it.

And no, unfortunately there's no `type.can_be_instantiated?` method.